### PR TITLE
rgbasm(5): Clarify charmap behavior (#558)

### LINKED ISSUE #558: `newcharmap` documentation
The documentation for `newcharmap` states that it creates a new, _empty_ charmap. However, this charmap is most certainly not empty:

```
	newcharmap test
	charmap "a", 1

SECTION "Test", ROM0[0]
Test:
	db "b"
```

Ideally, there should be a way to create an actually empty charmap, so that the above gives an error. But if this is not possible, at least the documentation should be updated.

### PR DESCRIPTION
**Pull request by dannye**
Fixes #558

### CODE REVIEW

**Comment by ISSOtm** [comment]
I would keep that charmaps, including the `main` one, are by default empty; and I would rather say that when no translation is performed, characters are copied from the source file verbatim.

Though I'm not entirely sure about the wording, this should be the gist of it.

**Comment by Rangi42** [comment]
Changing "default" to "empty" sounds good.

I would be confused by "copied from the source file verbatim" without already knowing about character encodings. How about "Any characters in a string without defined mappings will be copied directly, with the source file's encoding of characters to bytes."

**Comment by dannye** [comment]
Updated. Also, I simply put the new note at the bottom of the existing notes, but if you feel strongly about the notes being in a particular logical order, I can move them around.

**Comment by ISSOtm** [comment]
Hmm, re-reading it, I think it might still be confusing.

Should we briefly try to explain what an "encoding" is? Given that part of our user base is new to low-level, I think we should try to ease them into such concepts.

While we're at this, I'm noticing possibly outdated documentation in the surrounding area...

- [ ] Line 394:
    > When writing text that is meant to be displayed in the Game Boy, the characters used in the source code may have a different encoding than the default of ASCII.

    ASCII isn't really the default, especially given this PR...
    See below for a discussion of how encoding matters in RGBASM
- [ ] Line 397:
   > Character maps allow mapping strings up to 16 characters long to an abitrary 8-bit value:

   - [ ] There's a typo there!
   - [ ] Further, is the 16-char limit still a thing?
   - [ ] Should we replace "strings" with "character sequences" / "sequences of characters"?
   - [ ] Then there's an example of how to create mappings, but no example of what values would be output by `db "Amen<LF>"`
- [ ] Line 403:
    > By default, a character map contains ASCII encoding.
    
    Well, not quite, right? Again, see below.
    I think that this PR should modify this line instead of adding a footnote, and that an example (`db Amen<LF> ; Equivalent to db 128, 109, 101, 110, 10`) added below. 
- [ ] Line 421:
    > Character maps affect all strings in the file from the point in which they are defined, until switching to a different character map.
    
    Actually, this isn't true anymore! A charmap must be switched to explicitly. I believe that this is just poor wording to note that *modifications* to a charmap, i.e. new mappings, take effect immediately. I think that this line should be reworded entirely.
- [ ] Line 422:
    > This means that any string that the code may want to print as debug information will also be affected by it.
    
    This somewhat calls back to the example given before, but I think its wording is more confusing in this way. Plus, it's less relevant now, given that it's possible to switch between separate charmaps.
- [ ] Lines 425-426:
    > The output value of a mapping can be 0.
    > If this happens, the assembler will treat this as the end of the string and the rest of it will be trimmed.
    
    Is this still true?

# RGBASM input encoding

I just want to be sure about this... as far as I'm aware, RGBASM (and RGBDS as a whole, actually) are set up to accept ASCII and supersets, given that the C compilers we use encode all our strings as UTF-8. We don't actually rely on UTF-8, since nothing in the language itself accepts non-ASCII characters. The most that we do, is that upon encountering a non-ASCII character, we try to read an UTF-8 codepoint.

This is less true for strings; some special characters are still checked for (backslashes, newlines, obviously quotes, etc.), but they're mostly copied verbatim. I believe even invalid UTF-8 would be copied, actually... The charmap mappings themselves are only stored with bytes as indexes, not characters. In short, bytes are handled entirely transparently, so as long as the encoding is consistent across the document (or across document*s*, if `INCLUDE` came to be involved), RGBASM strings are encoding-agnostic.

**Comment by Rangi42** [comment]
@ISSOtm Your suggestions to @dannye look mostly fine. Clarifying some points:

> Should we briefly try to explain what an "encoding" is? Given that part of our user base is new to low-level, I think we should try to ease them into such concepts.

IMO not more than a couple sentences. How about: "There are thousands of text characters that can be stored in a file by encoding them as byte sequences. A character encoding such as UTF-8 defines the mapping from byte sequences to characters." If they need more info, that gives them the right search keywords.

> Further, is the 16-char limit still a thing?

No; `charmap "<abcdefghijklmnopqrstuvwxyz>", $42` works just fine.

> Should we replace "strings" with "character sequences" / "sequences of characters"?

I don't think so; in `charmap "<PK>", $12`, the `"<PK>"` is a standard quoted string literal.

> > The output value of a mapping can be 0.
> > If this happens, the assembler will treat this as the end of the string and the rest of it will be trimmed.
>
> Is this still true?

Yes.

**Comment by aaaaaa123456789** [comment]
I don't think RGBDS should explain encodings (beyond what's necessary to understand charmaps), but "further information" links (perhaps to wiki pages in the RGBDS repo) could always be a possibility. The documentation shouldn't grow to book-size just to contain a full copy of Assembly for Dummies, but there are no such limitations on the repo's wiki.

**Comment by dannye** [comment]
I reviewed the lines that @ISSOtm pointed out and made a few other tweaks that seemed appropriate.

I assume there is a more preferable way to write ``This would result in `db "Amen<LF>"` being equivalent to `db 128, 109, 101, 110, 10`.`` but I'll let you be the judge of that.

**Comment by Rangi42** [comment]
Regarding charmaps to NUL:

```
SECTION "test", ROM0

	charmap "<NULL>", $00
	charmap "<EXCL>", 33

	println "AB<NULL>CD<EXCL>?"
	db "AB<NULL>CD<EXCL>?"

S EQUS "EF<NULL>GH<EXCL>?"
	println "{S}"
	db "{S}"
```

This prints:

```
AB<NULL>CD<EXCL>?
EF<NULL>GH<EXCL>?
```

And it outputs:

```
41 42 00 43 44 21 3f 45 46 00 47 48 21 3f       AB\x00CD!?EF\x00GH!?
```

So NUL charmaps are okay; but it should note that charmaps don't apply to printing, only to ROM data output.

File: src/asm/rgbasm.5 [hunk 1]
```diff
@@ -391,16 +391,21 @@ with its corresponding argument in
...
-Character maps allow mapping strings up to 16 characters long to an abitrary 8-bit value:
+Character maps allow mapping strings to arbitrary 8-bit values:
 .Bd -literal -offset indent
 CHARMAP "<LF>", 10
 CHARMAP "&iacute", 20
 CHARMAP "A", 128
 .Ed
-By default, a character map contains ASCII encoding.
+This would result in
+.Sq db \(dqAmen<LF>\(dq
```
Comment (ISSOtm) [inline src/asm/rgbasm.5:404]: For code, `.Ql` is preferred. Otherwise, this is good!

**Review by ISSOtm — CHANGES_REQUESTED** [review]

**Review by ISSOtm — APPROVED** [review]

**Comment by ISSOtm** [comment]
0.5.0 time! :tada:

**Pull request merged**